control-plane/csi-driver/src/bin/node/fsfreeze/bin/mod.rs

```diff
@@ -0,0 +1,90 @@
+use crate::{
+    dev::Device,
+    error::FsfreezeError,
+    findmnt,
+    fsfreeze::bin::ioctl::{fsfreeze_ioctl, fsfreeze_preflight_check},
+    mount,
+};
+use strum_macros::{AsRefStr, Display, EnumString};
+use tracing::debug;
+use uuid::Uuid;
+
+pub(crate) mod ioctl;
+
+#[derive(EnumString, Clone, Debug, Eq, PartialEq, Display, AsRefStr)]
+#[strum(serialize_all = "lowercase")]
+pub enum FsFreezeOpt {
+    #[strum(serialize = "fs-freeze")]
+    Freeze,
+    #[strum(serialize = "fs-unfreeze")]
+    Unfreeze,
+}
+
+/// Issue FIFREEZE and FITHAW ioctl on the mount_path derived from the volume_id based on the
+/// command.
+pub(crate) async fn fsfreeze(volume_id: &str, command: FsFreezeOpt) -> Result<(), FsfreezeError> {
+    let uuid = Uuid::parse_str(volume_id).map_err(|error| FsfreezeError::InvalidVolumeId {
+        source: error,
+        volume_id: volume_id.to_string(),
+    })?;
+
+    let Some(device) =
+        Device::lookup(&uuid)
+            .await
+            .map_err(|error| FsfreezeError::InternalFailure {
+                source: error,
+                volume_id: volume_id.to_string(),
+            })?
+    else {
+        return Err(FsfreezeError::VolumeNotFound {
+            volume_id: volume_id.to_string(),
+        });
+    };
+
+    let device_path = device.devname();
+    if let Some(mnt) = mount::find_mount(Some(&device_path), None) {
+        // Make a preflight check, to ensure subsystem has at least one live path.
```

Comment:
wonder if we be a bit paranoid, and ensure only 1 path, and live ofcourse.

Reply:
They can be multiple live paths as well right, the duplicate thing that we saw the other day :-p ? 

But should be fail the freeze even if we have multiple live paths?

Reply:
how did we get duplicate live paths again?

Reply:
By adding a `duplicate` param to the nvme connect. Not sure if doing duplicate connect has any purpose.

Reply:
@tiagolobocastro Should I make this change or is it fine for now?

`
Make a preflight check, to ensure subsystem has at exactly one live path. Can't be 0 can't be more than 1.
`

Reply:
```
let live_subsystems: Vec<nvmeadm::nvmf_subsystem::Subsystem> =
        nvmeadm::nvmf_subsystem::Subsystem::try_from_nqn(nqn)
            .map_err(|_| Errno::ENODATA)?
            .into_iter()
            .filter(|subsystem| subsystem.state == "live")
            .collect();
    match live_subsystems.len() {
        0 => {
            error!(
                volume.id = volume_id,
                subsystem.nqn = nqn,
                "Preflight check for fsfreeze failed, subsystem has no live path"
            );
            Err(Errno::EHOSTUNREACH)
        }
        1 => Ok(()),
        _ => {
            error!(
                volume.id = volume_id,
                subsystem.nqn = nqn,
                "Preflight check for fsfreeze failed, subsystem has more than one live path"
            );
            Err(Errno::EMLINK)
        }
    }
```

Reply:
> By adding a `duplicate` param to the nvme connect. Not sure if doing duplicate connect has any purpose.

Ah right.. that was a "bug" though right? Though I guess we can simply use .all and ensure all paths are live then?